### PR TITLE
Feature: Layout

### PR DESCRIPTION
**Pull request by zero0Halo**
feature(page/layout): updated the existing page component to serve as layout 

feature(info): enabled storybook info add on to help auto-generate documentation

**Before submitting a pull request,** please make sure the following is done:

I have done **all** of the following:

- [x] Added a top level class to all my components `'.anchor-[COMPONENT NAME]'`.
- [ ] Used [conventional commits](https://www.conventionalcommits.org) for all work.
- [x] Tested my solution on Mobile & Tablet.
- [x] Wrote [unit tests](https://jestjs.io/docs/en/getting-started) for states and all behavior (`npm test`) and passed coverage thresholds.
- [x] Updated snapshots for all permutations (`npm test -- -u`).
- [ ] Accounted for hover, focus, blur, visited, & error states because they are not edge cases.
- [x] Created TODOs for known edge cases.
- [x] Documented all of my changes (inline & doc site).
- [ ] Made sure that all accessibility errors are resolved.
- [x] Added [stories](https://storybook.js.org/docs/basics/introduction/) with knobs for all possible configurations.
- [x] De-linted and ran [prettier](https://github.com/prettier/prettier) (`npm run pretty`) on my code.

---------
**Outline your feature or bug-fix below**

My code does...

### CODE REVIEW

**Comment by sharkparty** [comment]
checks are failing on this. Can we resync this with master and get the CI to pass? I'll still review it, but I imagine it'll be changing as I review...

File: .storybook/config.js [hunk 1]
```diff
@@ -1,10 +1,14 @@
 // STORYBOOK
 import { configure, addDecorator } from '@storybook/react';
 // ADDONS
+import { withInfo } from '@storybook/addon-info';
 import { withA11y } from '@storybook/addon-a11y';
 import { withKnobs } from '@storybook/addon-knobs';
 import { withViewport } from '@storybook/addon-viewport';
 
+addDecorator(withInfo({
+  source:false,
```
Comment (sharkparty) [inline .storybook/config.js:12]: Super-excited about this!

File: src/Icon/Icon.component.spec.tsx [hunk 2]
```diff
@@ -4,26 +4,20 @@ import * as renderer from 'react-test-renderer';
...
 // ENZYME
-import { shallow, mount } from 'enzyme';
+import { mount } from 'enzyme';
 
 // TEST SETUP
 const subject = <Icon type="hamburger" />;
-const component = shallow(subject);
 
-describe('Component: Button', () => {
+describe('Component: Icon', () => {
```
Comment (sharkparty) [inline src/Icon/Icon.component.spec.tsx:12]: Thank you!

File: src/Layout/Page/Page.component.tsx [hunk 3]
```diff
@@ -2,50 +2,215 @@
...
+    margin: 0 auto;
 `;
 
-const StyledBodyElement = styled.section`
-    flex: 1 0 auto;
-    background-color: ${colors.silver.base};
-    max-width: 100vw;
-    align-self: stretch;
-    padding: ${sizes.padding.lg} ${sizes.padding.xl};
+const StyledContent = styled.div<StyledContentProps>`
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx]: We may wanna use a `main` here in place of a `div` for SEO and best practices: https://www.w3schools.com/tags/tag_main.asp

File: src/Layout/Page/Page.component.tsx [hunk 4]
```diff
@@ -2,50 +2,215 @@
...
+    max-width: ${props => props.contentWidth};
+    background-color: ${props => props.contentBackgroundColor};
+    margin: 0 auto;
 `;
 
+// Interfaces
+// ------------------------------------------------------------------------------------------------------------------
 interface PageProps {
+    /** The header of the website. This can be a component. */
+    header?: any;
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx:29]: I don't know the answer to this question, but is there a way we can enforce only the right element types AND use `children`? My only hack to finding out how to do this is to type-cast the props (which works with what you have here), but I'd like to be able to pass all these elements in as children and have them all work 'automatically'.

File: src/Layout/Page/Page.component.tsx [hunk 5]
```diff
@@ -2,50 +2,215 @@
...
+    /** The width of the content within the layout. */
+    contentWidth?: string | number;
+    /** The width of the area for the sidebar. */
+    sidebarWidth?: string | number;
+    /** If the sidebar is on the left or the right of the layout */
+    sidebarAlign?: SidebarAlignment;
+    /** The actual sidebar. This can be a component. */
+    sidebar?: any;
+    /** The background color of the layout. Default is transparent. */
+    layoutBackgroundColor?: string;
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx]: This prop and the subsequent one can be a slippery slope. I'm kosher leaving them here, but the second I start hearing about needing font-color, anchor tag treatment, etc. overrides, I'll yank these props out of here and ask people to extend this using classes.

File: src/Layout/Page/Page.component.tsx [hunk 6]
```diff
@@ -2,50 +2,215 @@
...
+    /** The background color of the content within the layout area. Default is transparent. */
+    contentBackgroundColor?: string;
+    /** Additional classname. */
+    className?: string;
+    children?: any;
+}
+
+interface SidebarLayoutProps {
+    /** The width of the layout, edge to edge. */
+    layoutWidth?: string | number;
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx]: Do are you policing the units here? If you allow only a number then everything can always be in REMs, but it's not quite as intuitive for downstream users. If you allow strings, then people can pass in anything and that can get unwieldy too. I'm not positing a preference, here, but more asking what your thoughts are on this?

File: src/Layout/Page/Page.component.tsx [hunk 7]
```diff
@@ -2,50 +2,215 @@
...
 }
 
-export const Page = ({ children }: PageProps): React.ReactElement<any> => (
-    <StyledPageElement className="anchor-page">
+interface StyledLayoutProps {
+    layoutWidth?: string | number;
+    layoutBackgroundColor?: string;
+}
+
+interface StyledContentProps {
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx]: Can we break all of these components apart?

File: src/Layout/Page/Page.component.tsx [hunk 8]
```diff
@@ -2,50 +2,215 @@
...
+    contentWidth?: string | number;
+    contentBackgroundColor?: string;
+}
+
+// Components
+// ------------------------------------------------------------------------------------------------------------------
+export const Page = (props: PageProps): React.ReactElement<any> => (
+    <StyledPage className={classnames('anchor-page', props.className)}>
         <NormalizeCSS />
         <GlobalCSS />
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx]: I'm debating on having `GlobalCSS`. What are your thoughts on this?

File: src/Layout/Page/Page.component.tsx [hunk 9]
```diff
@@ -2,50 +2,215 @@
...
-        <Header />
-        <StyledBodyElement>{children}</StyledBodyElement>
-        <Footer />
-    </StyledPageElement>
+        <Grid
+            columns={1}
+            rows={`minmax(${props.headerHeight},auto) 1fr minmax(${
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx]: Did this pass the prettier/linter?

File: src/Layout/Page/Page.component.tsx [hunk 10]
```diff
@@ -2,50 +2,215 @@
...
+            {props.header && <Cell>{props.header}</Cell>}
+
+            <Cell>{props.children}</Cell>
+
+            {props.footer && <Cell>{props.footer}</Cell>}
+        </Grid>
+    </StyledPage>
+);
+
+Page.defaultProps = PageDefaultProps;
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx]: Let's talk about default props. I think we need a standard here and I wanna hear your input. I'd either like all defaults to be in the method signature, or all defaults like this. My opinion is that this isn't quite as readable for the average person, but I do wanna chat about it from other perspectives.

File: src/Layout/Page/Page.component.tsx [hunk 11]
```diff
@@ -2,50 +2,215 @@
...
+
+const DefaultLayout = (
+    props: DefaultLayoutProps
+): React.ReactElement<any> => (
+    <StyledLayout
+        layoutWidth={props.layoutWidth}
+        layoutBackgroundColor={props.layoutBackgroundColor}
+        className={classnames('anchor-default-layout', props.className)}
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx]: Does this need to be `default-layout` or can it just be layout? It's our only layout so default is kind of implied IMO.

File: src/Layout/Page/Page.component.tsx [hunk 12]
```diff
@@ -2,50 +2,215 @@
...
+    <StyledLayout
+        layoutWidth={props.layoutWidth}
+        layoutBackgroundColor={props.layoutBackgroundColor}
+        className={classnames('anchor-sidebar-layout', props.className)}
+    >
+        <StyledContent
+            contentWidth={props.contentWidth}
+            contentBackgroundColor={props.contentBackgroundColor}
+        >
+            {props.sidebarAlign === RIGHT ? (
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx]: Ahh, typescript shows a little value here: we can use strict props to make binary configuration out of strings without having to have messy conditional blocks that check both sides of the condition.

File: src/Layout/Page/Page.component.tsx [hunk 13]
```diff
@@ -2,50 +2,215 @@
...
+}
+
+interface StyledContentProps {
+    contentWidth?: string | number;
+    contentBackgroundColor?: string;
+}
+
+// Components
+// ------------------------------------------------------------------------------------------------------------------
+export const Page = (props: PageProps): React.ReactElement<any> => (
```
Comment (sharkparty) [inline src/Layout/Page/Page.component.tsx]: Can we destructure all props?

File: src/Layout/Page/Page.spec.tsx [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+  // DEFAULT_SIDEBAR_WIDTH,
+  // TRANSPARENT,
+} from './Page.component';
+
+describe('Component: Page', () => {
+  const subject = <PageComponent />;
+  const wrapper = mount(subject);
+  const component = shallow(subject);
+
+  it('should be defined.', () => {
```
Comment (sharkparty) [inline src/Layout/Page/Page.spec.tsx]: `'should match it's default snapshot.'`

File: src/Layout/Page/Page.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+  DEFAULT_HEADER_HEIGHT,
+  DEFAULT_FOOTER_HEIGHT,
+  DEFAULT_LAYOUT_WIDTH,
+  TRANSPARENT,
+  DEFAULT_CONTENT_WIDTH,
+  // DEFAULT_CONTENT_WIDTH,
+  // DEFAULT_SIDEBAR_WIDTH,
+  // TRANSPARENT,
+} from './Page.component';
+
```
Comment (sharkparty) [inline src/Layout/Page/Page.spec.tsx:11]: `const {  DefaultLayout  } = Page;` can make things easier.

File: src/Layout/Page/Page.stories.tsx [hunk 16]
```diff
@@ -3,13 +3,80 @@ import * as React from 'react';
...
+    background: ${colors.white.base};
+`;
+
+// Stories
+// ------------------------------------------------------------------------------------------------------------------
+storiesOf('Components/Layout/Page', module).add('Default Layout', () => (
+    <StyledStory>
+        <PageComponent header={<Header />} footer={<Footer />}>
+            <Page.DefaultLayout
+                layoutWidth={text('Layout Width', DEFAULT_LAYOUT_WIDTH)}
```
Comment (sharkparty) [inline src/Layout/Page/Page.stories.tsx]: This is so neat.

File: src/theme/GlobalStyles/Globals.styles.ts [hunk 17]
```diff
@@ -22,6 +22,7 @@ export const GlobalCSS = createGlobalStyle`
     overflow-y: hidden;
     overflow-x: hidden;
     height: 100vh;
+    /* TODO: Change this back to #ffffff */
     background-color: #710a32;
```
Comment (sharkparty) [inline src/theme/GlobalStyles/Globals.styles.ts:26]: Thanks!

**Review by sharkparty — CHANGES_REQUESTED** [review]
Added a few comments to get the discussion going. HMU with questions or comments.

**Comment by sharkparty** [comment]
Wow, github is really scary with its changes-requested treatment. That's intimidating and I don't like it.

File: src/Layout/Page/Page.component.tsx (hunk 3 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx]: I think this is a good idea. I wasn't crazy about using `div` or `section` for the main wrapper. Ironically the `main` tag was looking me right in the face and I didn't even think of it.

File: src/Layout/Page/Page.component.tsx (hunk 4 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx:29]: Using props is more standard/natural way of approaching the problem, and I think trying to not be too clever on this is the best way to go. Still, the idea of sniffing out a child's type and applying it correctly is an interesting notion. Definitely would like to talk about this because it's intriguing.

File: src/Layout/Page/Page.component.tsx (hunk 5 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx]: Agreed. I put the layout/content background color values for common use cases I was seeing on RMN's site where the background colors were extending from edge to edge, but the content was in the center. Granted that could be taken care of solely in the user's own component, but it got me thinking that perhaps giving them additional options might be good too. 

But yeah, it does potentially open a door for an onslaught of options.

File: src/Layout/Page/Page.component.tsx (hunk 6 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx]: I had been questioning having string/number as the type almost from the get-go. I think I'm going to yank `number` and keep `string` because ultimately that allows for more flexibility and less confusion.

File: src/Layout/Page/Page.component.tsx (hunk 7 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx]: That's my plan. I'm going to do it in an upcoming push.

File: src/Layout/Page/Page.component.tsx (hunk 8 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx]: I did most of the work on this w/o it, so I'm fine with ditching it. One possibility, a prop to enable it?

File: src/Layout/Page/Page.component.tsx (hunk 9 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx]: Yep!

File: src/Layout/Page/Page.component.tsx (hunk 10 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx]: I'm up for the discussion, I'm definitely not married to the way I have it.

File: src/Layout/Page/Page.component.tsx (hunk 11 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx]: There's two layouts, DefaultLayout and SidebarLayout. I went this route because this naming gives us latitude if we wanted to add more layouts. I think the door should be left open for that possibility, because although it's all we have right now that doesn't mean the list won't get bigger.

File: src/Layout/Page/Page.component.tsx (hunk 12 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx]: I'd like to talk about this more as I'm not sure how to go about that.

File: src/Layout/Page/Page.component.tsx (hunk 13 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.component.tsx]: Yeah that should be fine. Originally I was spreading `props` to the styled components, but I stopped doing that.

File: src/Layout/Page/Page.spec.tsx (hunk 15 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.spec.tsx:11]: Since I'm going to breaking up the components this shouldn't be an issue (soon).

File: src/Layout/Page/Page.stories.tsx (hunk 16 above)
Comment (zero0Halo) [reply src/Layout/Page/Page.stories.tsx]: Gotta love those knobs.

File: src/Layout/Page/Page.component.tsx (hunk 3 above)
Comment (sharkparty) [reply src/Layout/Page/Page.component.tsx]: Yeah, I often lose sight of the trees in the forrest. I'm pretty sure the content area for Rx is inside of a `section` 🙈

File: src/Layout/Page/Page.component.tsx (hunk 4 above)
Comment (sharkparty) [reply src/Layout/Page/Page.component.tsx:29]: Maybe we add a v2 ticket. This can be outside of MVP, but still worth considering, right?

File: src/Layout/Page/Page.component.tsx (hunk 5 above)
Comment (sharkparty) [reply src/Layout/Page/Page.component.tsx]: For some of those we can create a `Hero` component that functions like Bootstrap's Hero, but this also solves that nicely. I just wanna get ahead the 'onslaught', but I don't think that merits a change to this. It's a nice feature to have.

File: src/Layout/Page/Page.component.tsx (hunk 6 above)
Comment (sharkparty) [reply src/Layout/Page/Page.component.tsx]: SGTM. I did have Icons implemented the other way, but IIRC Icons now take in a scale prop that's an Enum. Can you do me a favor and check if `Typography` is consistent with this pattern? I think it only takes in a `number` and it assumed that that number is in REMs. If so, can you create a bug ticket and assign it to me to make that consistent with this approach?

File: src/Layout/Page/Page.component.tsx (hunk 8 above)
Comment (sharkparty) [reply src/Layout/Page/Page.component.tsx]: I don't think it'll define anything anymore TBH. I created a Typography component that renders all permutations of our type, so we no longer need to have typography in there and that's all that was there. The Normalize is probably the only thing we need and IDK if that will be essential as we build out components with known un-normalized CSS we can just build that into the component.

File: src/Layout/Page/Page.component.tsx (hunk 10 above)
Comment (sharkparty) [reply src/Layout/Page/Page.component.tsx]: I'll add a mtg to discuss this. I'll also try and get a few others in the room to talk about it.

File: src/Layout/Layout.component.tsx [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+    className?: string;
+    children?: any;
+}
+
+interface StyledLayoutProps {
+    layoutWidth?: string;
+    layoutBackgroundColor?: string;
+}
+
+interface StyledContentProps {
```
Comment (sharkparty) [inline src/Layout/Layout.component.tsx:51]: Do we want to extricate `Content` from `Layout` as a nested component in a nested dir?

File: src/Layout/Layout.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+        expect(testSubject.find('StyledContent').prop('contentWidth')).toEqual(
+            contentWidth
+        );
+        expect(
+            testSubject.find('StyledContent').prop('contentBackgroundColor')
+        ).toEqual(contentBackgroundColor);
+    });
+});
+
+// tslint:disable-next-line: no-console
```
Comment (sharkparty) [inline src/Layout/Layout.spec.tsx]: This lint disabler and the console.log can be removed.

File: src/Layout/Layout.stories.tsx [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+                )}
+            >
+                <h1>Default Layout</h1>
+                <h2>This example uses the Page component as a wrapper.</h2>
+            </Layout>
+        </Page>
+    </StyledStory>
+));
+
+// storiesOf('Components/Layout/Page', module)
```
Comment (sharkparty) [inline src/Layout/Layout.stories.tsx]: If we don't want these stories, then let's delete them. Add the readme scaffold, though. That we'll likely want downstream.

File: src/Layout/Page/Page.stories.tsx [hunk 21]
```diff
@@ -3,21 +3,50 @@ import * as React from 'react';
...
+import {
+    Page,
+    DEFAULT_HEADER_HEIGHT,
+    DEFAULT_FOOTER_HEIGHT,
+} from './Page.component';
+import { Footer } from '../Footer/Footer.component';
+import { Header } from '../Header/Header.component';
+import { colors } from '../../theme';
 // README
-import * as README from './README.md';
```
Comment (sharkparty) [inline src/Layout/Page/Page.stories.tsx]: We'll prolly wanna keep the readme, but lose the other commented out code.

File: src/Layout/SidebarLayout/SidebarLayout.component.tsx [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+interface StyledContentProps {
+    contentWidth?: string;
+    contentBackgroundColor?: string;
+}
+
+// Components
+// ------------------------------------------------------------------------------------------------------------------
+export const SidebarLayout = ({
```
Comment (sharkparty) [inline src/Layout/SidebarLayout/SidebarLayout.component.tsx:74]: Finest possible hair being split here: Do we need the added `Layout` for this component name? It sounds redundant, but I will definitely not die on this hill.

Comment (zero0Halo) [reply src/Layout/SidebarLayout/SidebarLayout.component.tsx:74]: I think w/o the word 'Layout' the usage of the component could be confusing. Dropping `<SidebarLayout>` in jsx is much more clear than `Sidebar` IMO

File: src/Layout/Layout.stories.tsx (hunk 20 above)
Comment (zero0Halo) [reply src/Layout/Layout.stories.tsx]: That was me being dumb. Never meant to leave the commented garbage.

File: src/Layout/Layout.component.tsx (hunk 18 above)
Comment (zero0Halo) [reply src/Layout/Layout.component.tsx:51]: The reason I have Layout and Content styled components is to more easily be able to change the width of both of them via the single Layout (or SidebarLayout) component(s). I don't think it's necessary (right now anyways) to make them individual components for the user to consume. Breaking them up like that removes some of the simplicity of just dropping in `Layout` or `SidebarLayout`. I'm open to discuss this more though, as perhaps I'm not seeing the same angle as you.

Comment (sharkparty) [reply src/Layout/Layout.component.tsx:51]: 👍

**Review by sharkparty — APPROVED** [review]
LGTM
